### DISCUSSION TITLE
Not include the lib in each class we want to serialize

### DISCUSSION THREAD

**Discussion by Tudyx** [nlohmann/json#2605] [Q&A] [answered]
Could you please advise me on something. I want to serialize multiple classes which all have private members.  
Reading the doc it's seem that the traditional solution is to include the lib in each classes and use `to_json` and `from_json` or the `macro` for the serialization.  
Since the size of the lib is quite big,  i have another option but i am not sure if it's worth it. The solution is to include a struct in all of this classes and declare it as a friend. The header of this struct will not include the lib, only the `.cpp` file will. The methods will contain `void` pointers which will be caste to JSON reference in the `.cpp` file to do the serialization thing.  
This is quite tricky and not trivial, do you think it's worth it just for not include the lib in each class ?

**Comment by nlohmann** [comment, marked as answer]
There is a forward-declaration header in the multi-header version (see `include` path).

Then such code could work:

```cpp
// foo.hpp

#include <nlohmann/json_fwd.hpp>

class foo {
  private:
	int m_i = 0;

  public:
  	foo(int i);

  	friend void to_json(nlohmann::json& j, const foo& f);
};
```

```cpp
// foo.cpp

#include "foo.hpp"
#include <nlohmann/json.hpp>
#include <iostream>

foo::foo(int i)
: m_i(i)
{}

void to_json(nlohmann::json& j, const foo& f)
{
	j["i"] = f.m_i;
}

int main()
{
	foo f(17);
	nlohmann::json j = f;
	std::cout << j << std::endl;
}
```

**Comment by Tudyx** [comment]
Totally solved my problem, thanks a lot!
